File: assets/scripts/core/StarNetworkModel.ts
// M02《点亮你温暖我》核心机制模型 —— 引擎无关的纯逻辑, 由 Vitest 钉死正确性.
// 规则(spec §5.3): 点一颗星=该星+全部直连邻居满命(lifeMax); 每拍全体同时衰减:
//   亮邻居数 >= freezeThreshold 则冻结(不掉命), 否则 -1, 归0熄灭.
//   胜利 = 所有星都亮 且 每颗亮邻居 >= freezeThreshold (整网自稳锁死).
// 语义与 scripts/m02-starweb-verify.mjs 的求解器逐位对齐(衰减用结算前快照).

export interface StarNetworkRules {
  lifeMax: number;         // 拍；星被点亮后的满命
  freezeThreshold: number; // 个；冻结所需亮邻居数
}

export interface BoardGraph {
  nodes: string[];
  edges: [string, string][];
}

export class StarNetworkModel {
  private readonly rules: StarNetworkRules;
  private readonly nodes: string[];
  private readonly neighbors: Map<string, string[]>;
  private life: Map<string, number>;

  constructor(graph: BoardGraph, rules: StarNetworkRules) {
    this.rules = rules;
    this.nodes = [...graph.nodes];
    this.neighbors = new Map(this.nodes.map((n) => [n, [] as string[]]));
    for (const [a, b] of graph.edges) {
      this.neighbors.get(a)?.push(b);
      this.neighbors.get(b)?.push(a);
    }
    this.life = new Map(this.nodes.map((n) => [n, 0]));
  }

  lifeOf(id: string): number {
    return this.life.get(id) ?? 0;
  }

  isLit(id: string): boolean {
    return this.lifeOf(id) > 0;
  }

  neighborsOf(id: string): readonly string[] {
    return this.neighbors.get(id) ?? [];
  }

  /** 当前状态下 id 的亮邻居数 */
  litNeighborCount(id: string): number {
    return this.countLitNeighbors(this.life, id);
  }

  /** 点亮 id 及其全部直连邻居到满命 (未知星忽略) */
  tap(id: string): void {
    if (!this.life.has(id)) return;
    this.life.set(id, this.rules.lifeMax);
    for (const n of this.neighborsOf(id)) {
      this.life.set(n, this.rules.lifeMax);
    }
  }

  /** 全体同时衰减一拍: 用结算前快照判定, 亮邻居不足 freeze 的亮星 -1 */
  tick(): void {
    const snapshot = new Map(this.life);
    for (const id of this.nodes) {
      const life = snapshot.get(id) ?? 0;
      if (life <= 0) continue;
      if (this.countLitNeighbors(snapshot, id) < this.rules.freezeThreshold) {
        this.life.set(id, life - 1);
      }
    }
  }

  /** 一拍 = 点亮 + 衰减 */
  step(id: string): void {
    this.tap(id);
    this.tick();
  }

  /** 胜利: 全亮且每颗亮邻居 >= freezeThreshold */
  isWon(): boolean {
    return this.nodes.every(
      (id) => this.isLit(id) && this.litNeighborCount(id) >= this.rules.freezeThreshold
    );
  }

  reset(): void {
    this.life = new Map(this.nodes.map((n) => [n, 0]));
  }

  private countLitNeighbors(state: Map<string, number>, id: string): number {
    let count = 0;
    for (const n of this.neighborsOf(id)) {
      if ((state.get(n) ?? 0) > 0) count += 1;
    }
    return count;
  }
}

File: StarGuardian/Assets/Scripts/Core/StarNetworkModel.cs
// M02《点亮你温暖我》核心机制模型 —— 引擎无关的纯逻辑, 由 xUnit 钉死正确性.
// 规则(spec §5.3): 点一颗星=该星+全部直连邻居满命(lifeMax); 每拍全体同时衰减:
//   亮邻居数 >= freezeThreshold 则冻结(不掉命), 否则 -1, 归0熄灭.
//   胜利 = 所有星都亮 且 每颗亮邻居 >= freezeThreshold (整网自稳锁死).
// 语义由 unity-tests 单测钉死(衰减用结算前快照); 从 assets/scripts/core/StarNetworkModel.ts 迁移, 规则不变.

using System.Collections.Generic;
using System.Linq;

namespace StarGuardian.Core
{
    public readonly struct StarNetworkRules
    {
        /// <summary>拍；星被点亮后的满命</summary>
        public int LifeMax { get; init; }

        /// <summary>个；冻结所需亮邻居数</summary>
        public int FreezeThreshold { get; init; }
    }

    public sealed class BoardGraph
    {
        public IReadOnlyList<string> Nodes { get; }
        public IReadOnlyList<(string A, string B)> Edges { get; }

        public BoardGraph(IEnumerable<string> nodes, IEnumerable<(string A, string B)> edges)
        {
            Nodes = nodes.ToList();
            Edges = edges.ToList();
        }
    }

    public sealed class StarNetworkModel
    {
        private readonly StarNetworkRules rules;
        private readonly List<string> nodes;
        private readonly Dictionary<string, List<string>> neighbors;
        private Dictionary<string, int> life;

        public StarNetworkModel(BoardGraph graph, StarNetworkRules rules)
        {
            this.rules = rules;
            nodes = new List<string>(graph.Nodes);
            // 用 Set 去重邻接: 镜像边 [A,B]+[B,A] 或重复边不得把邻居算两次(否则虚高的
            // litNeighborCount 会造成假冻结/假通关); 同时忽略自环与指向未知节点的边.
            var adjacency = nodes.ToDictionary(n => n, _ => new HashSet<string>());
            foreach (var (a, b) in graph.Edges)
            {
                if (a != b && adjacency.ContainsKey(a) && adjacency.ContainsKey(b))
                {
                    adjacency[a].Add(b);
                    adjacency[b].Add(a);
                }
            }
            neighbors = adjacency.ToDictionary(kv => kv.Key, kv => kv.Value.ToList());
            life = nodes.ToDictionary(n => n, _ => 0);
        }

        public int LifeOf(string id) => life.TryGetValue(id, out var v) ? v : 0;

        public bool IsLit(string id) => LifeOf(id) > 0;

        public IReadOnlyList<string> NeighborsOf(string id) =>
            neighbors.TryGetValue(id, out var list) ? list : (IReadOnlyList<string>)System.Array.Empty<string>();

        /// <summary>当前状态下 id 的亮邻居数</summary>
        public int LitNeighborCount(string id) => CountLitNeighbors(life, id);

        /// <summary>点亮 id 及其全部直连邻居到满命 (未知星忽略)</summary>
        public void Tap(string id)
        {
            if (!life.ContainsKey(id)) return;
            life[id] = rules.LifeMax;
            foreach (var n in NeighborsOf(id))
            {
                life[n] = rules.LifeMax;
            }
        }

        /// <summary>全体同时衰减一拍: 用结算前快照判定, 亮邻居不足 freeze 的亮星 -1</summary>
        public void Tick()
        {
            var snapshot = new Dictionary<string, int>(life);
            foreach (var id in nodes)
            {
                var current = snapshot.TryGetValue(id, out var v) ? v : 0;
                if (current <= 0) continue;
                if (CountLitNeighbors(snapshot, id) < rules.FreezeThreshold)
                {
                    life[id] = current - 1;
                }
            }
        }

        /// <summary>一拍 = 点亮 + 衰减。未知星不构成一次点亮 → 不推进衰减(不白耗一拍)，返回是否真发生了点亮</summary>
        public bool Step(string id)
        {
            if (!life.ContainsKey(id)) return false;
            Tap(id);
            Tick();
            return true;
        }

        /// <summary>胜利: 全亮且每颗亮邻居 >= freezeThreshold (空图不算胜利)</summary>
        public bool IsWon() =>
            nodes.Count > 0 &&
            nodes.All(id => IsLit(id) && LitNeighborCount(id) >= rules.FreezeThreshold);

        public void Reset()
        {
            life = nodes.ToDictionary(n => n, _ => 0);
        }

        private int CountLitNeighbors(Dictionary<string, int> state, string id)
        {
            var count = 0;
            foreach (var n in NeighborsOf(id))
            {
                if (state.TryGetValue(n, out var v) && v > 0) count += 1;
            }
            return count;
        }
    }
}
